### PR TITLE
wallet: Migrate legacy wallets to descriptor wallets without requiring BDB

### PR DESCRIPTION
**Pull request by achow101**
#26606 introduced `BerkeleyRODatabase` which is an independent parser for BDB files. This PR uses this in legacy wallet migration so that migration will continue to work once the legacy wallet and BDB are removed. `LegacyDataSPKM` is introduced to have the minimum data and functions necessary for a legacy wallet to be loaded for migration.

### CODE REVIEW

**Comment by Sjors** [comment]
I think it would be easier to review if you made a separate PR for `BerkeleyRODatabase` and have it work as a regular, albeit read-only, wallet database backend.

I'm pleasantly surprised at how small the implementation is.

**Comment by luke-jr** [comment]
I don't see how reimplementing BDB is better than just continuing to use BDB.

**Comment by achow101** [comment]
> I don't see how reimplementing BDB is better than just continuing to use BDB.

The reimplementation is quite compact because it is for only reading BDB files, only the features that we use, and none of the actual DB system stuff. BDB itself contains a ton of code, a lot of which we don't use. Maintaining it as a dependency will eventually be more burdensome - we already need to patch it in order to compile in the depends system. This reimplementation does not carry those issues - it implements a format that isn't going to change, and it only depends on things that are already being used elsewhere in the codebase.

**Comment by achow101** [comment]
> I think it would be easier to review if you made a separate PR for `BerkeleyRODatabase` and have it work as a regular, albeit read-only, wallet database backend.

I've opeend #26606 for just `BerkeleyRODatabase` and its use in wallettool's `dump`.

**Comment by achow101** [comment]
Rebased following #26606. Now ready for review.

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -4346,11 +4346,37 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
...
+        wallet = CWallet::Create(empty_context, wallet_name, std::move(database), options.create_flags, error, warnings);
+        if (!wallet) {
+            return util::Error{Untranslated("Wallet loading failed.") + Untranslated(" ") + error};
+        }
+
+        if (wallet->IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
+            return util::Error{_("Error: This wallet is already a descriptor wallet")};
+        }
+        // Unload
+        wallet.reset();
```
Comment (furszy) [inline src/wallet/wallet.cpp:4402]: Loading the entire wallet just to obtain the descriptors flag shouldn't be needed. Could just check the file type instead. We want a bdb file here. So, something like https://github.com/furszy/bitcoin-core/commit/7a03a039ae99e69504fe9df95b04f5e326c45a70 would work well. Yet, the code isn't the best but it includes a test that verifies the behavior.

Comment (achow101) [reply src/wallet/wallet.cpp]: Good point, I think the original intention was to be able to handle the legacy-sqlite case, but that's moot since we require BDB_RO later. I've changed this to check for just look for BDB files, and also added your test.

File: src/wallet/scriptpubkeyman.cpp [hunk 2]
```diff
@@ -1812,7 +1953,7 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
         WalletDescriptor w_desc(std::move(desc), creation_time, 0, 0, 0);
 
         // Make the DescriptorScriptPubKeyMan and get the scriptPubKeys
-        auto desc_spk_man = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(m_storage, w_desc, m_keypool_size));
+        auto desc_spk_man = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(m_storage, w_desc, /*keypool_size=*/0));
```
Comment (cbergqvist) [inline src/wallet/scriptpubkeyman.cpp:1956]: Worth adding a comment about the consequences of setting `keypool_size=0`? I'm guessing there are still cases where we generate some pubkeys despite having a zero pool size, otherwise you would have removed some of the code and the for-loop below?

Comment (cbergqvist) [inline src/wallet/scriptpubkeyman.cpp:1956]: nit: Might as well use `make_unique` on the lines you modify:
```suggestion
        auto desc_spk_man = std::make_unique<DescriptorScriptPubKeyMan>(m_storage, w_desc, /*keypool_size=*/0);
```

File: src/wallet/walletdb.cpp [hunk 3]
```diff
@@ -763,7 +763,7 @@ static DBErrors LoadLegacyWalletRecords(CWallet* pwallet, DatabaseBatch& batch,
 
         // nTimeFirstKey is only reliable if all keys have metadata
         if (pwallet->IsLegacy() && (key_res.m_records + ckey_res.m_records + watch_script_res.m_records) != (keymeta_res.m_records + watch_meta_res.m_records)) {
-            auto spk_man = pwallet->GetOrCreateLegacyScriptPubKeyMan();
+            auto spk_man = pwallet->GetLegacyScriptPubKeyMan();
```
Comment (cbergqvist) [inline src/wallet/walletdb.cpp:766]: Why not `OrCreate` here any longer?

**Review by cbergqvist — COMMENTED** [review]
Code reviewed e11fa4524723aebc0d879ee083c2d6d46849d89a.

Overall concept: Not fully convinced about the benefit of breaking out `LegacyDataSPKM`, seems to add complexity. Is the goal to delete `LegacyScriptPubKeyMan` in the future when ripping out BDB and only keep `LegacyDataSPKM`?

Why perform the inlining of `IsMine()` in 31e918afc3920a2f8aa367ad323f590b4856aa26 before the function is removed? Seems risky in case of other PRs modifying it prior to removal etc.

`LegacyScriptPubKeyMan::NewKeyPool()` only touches data from `LegacyDataSPKM`, should it be moved to that class as public (or protected and then exposed in the subclass?). (If `LegacyScriptPubKeyMan` is slated for removal I can understand you want to signal that method will no longer be needed). (Have not checked if there are other methods along the same lines).

Assuming `LegacyDataSPKM` name is kept, comment in `src/script/signingprovider.h` could be updated to refer to `LegacyDataSPKM::setWatchOnly` instead.

**Comment by achow101** [comment]
> Is the goal to delete `LegacyScriptPubKeyMan` in the future when ripping out BDB and only keep `LegacyDataSPKM`?

Yes.

We are trying to remove the legacy wallet in its entirety, leaving only behind the absolute minimum necessary to perform migration.

> Why perform the inlining of `IsMine()` in [31e918a](https://github.com/bitcoin/bitcoin/commit/31e918afc3920a2f8aa367ad323f590b4856aa26) before the function is removed? Seems risky in case of other PRs modifying it prior to removal etc.

`IsMine()` is unlikely to change. Besides the fact that any changes to the legacy wallet are likely to be closed instead of seriously considered, `IsMine()` has not had any material changes in the last several years. It could be included in #28710 but that PR is big enough as it is.

> `LegacyScriptPubKeyMan::NewKeyPool()` only touches data from `LegacyDataSPKM`, should it be moved to that class as public (or protected and then exposed in the subclass?)

No. The separation exists now to indicate which things can be removed at the end. `LegacyDataSPKM` is the bare minimum that must stick around for migration, everything else remains in `LegacyScriptPubKeyMan` and will be removed at the same time.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -4438,6 +4465,7 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
     // both before and after reloading. This ensures the set is complete even if one of the wallets
     // fails to reload.
     std::set<fs::path> wallet_dirs;
+    options.require_format = std::nullopt;
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/wallet.cpp:4487]: It's used by the `reload_wallet` calls which capture it.

File: src/wallet/scriptpubkeyman.h [hunk 5]
```diff
@@ -276,31 +276,106 @@ static const std::unordered_set<OutputType> LEGACY_OUTPUT_TYPES {
...
 
     //! Number of pre-generated keys/scripts (part of the look-ahead process, used to detect payments)
     int64_t m_keypool_size GUARDED_BY(cs_KeyStore){DEFAULT_KEYPOOL_SIZE};
 
-    bool AddKeyPubKeyInner(const CKey& key, const CPubKey &pubkey);
-    bool AddCryptedKeyInner(const CPubKey &vchPubKey, const std::vector<unsigned char> &vchCryptedSecret);
+    bool AddKeyPubKeyInner(const CKey& key, const CPubKey &pubkey) override;
 
     /**
      * Private version of AddWatchOnly method which does not accept a
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/scriptpubkeyman.h:381]: I don't think it's necessary to add changes that correct this documentation when it's all going to be deleted soon anyways.

File: src/wallet/wallet.h [hunk 6]
```diff
@@ -961,6 +961,8 @@ class CWallet final : public WalletStorage, public interfaces::Chain::Notificati
     //! Get the LegacyScriptPubKeyMan which is used for all types, internal, and external.
     LegacyScriptPubKeyMan* GetLegacyScriptPubKeyMan() const;
     LegacyScriptPubKeyMan* GetOrCreateLegacyScriptPubKeyMan();
+    LegacyDataSPKM* GetLegacyDataSPKM() const;
+    LegacyDataSPKM* GetOrCreateLegacyDataSPKM();
 
     //! Make a LegacyScriptPubKeyMan and set it for all types, internal, and external.
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/wallet.h:967]: The function still creates a `LegacyScriptPubKeyMan` in normal operation.

File: src/wallet/scriptpubkeyman.cpp [hunk 7]
```diff
@@ -1887,7 +1886,6 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
             for (const CScript& spk : desc_spks) {
                 size_t erased = spks.erase(spk);
                 assert(erased == 1);
-                assert(IsMine(spk) == ISMINE_SPENDABLE);
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp:1869]: No. `LegacyDataSPKM` does not have an `IsMine()` function, so this call needs to be removed otherwise it will not compile.

File: src/wallet/scriptpubkeyman.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp:1956]: I don't think this needs to have any additional explanation. `keypool_size` only matters for ranged descriptors. Non-HD keys do not make ranged descriptors, so the value here is ignored. The loop below is for an entirely different set of descriptors and is unrelated to this.

File: src/wallet/walletdb.cpp (hunk 3 above)
Comment (achow101) [reply src/wallet/walletdb.cpp:766]: It's redundant since `IsLegacy()` ensures the existence of a `LegacyScriptPubKeyMan`, so there's no need to try to create one.

File: src/wallet/scriptpubkeyman.h [hunk 8]
```diff
@@ -276,31 +276,106 @@ static const std::unordered_set<OutputType> LEGACY_OUTPUT_TYPES {
 
 class DescriptorScriptPubKeyMan;
 
-class LegacyScriptPubKeyMan : public ScriptPubKeyMan, public FillableSigningProvider
+class LegacyDataSPKM : public ScriptPubKeyMan, public FillableSigningProvider
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/scriptpubkeyman.h:281]: Add a comment to each

File: src/wallet/walletdb.cpp [hunk 9]
```diff
@@ -354,7 +354,7 @@ bool LoadKey(CWallet* pwallet, DataStream& ssKey, DataStream& ssValue, std::stri
             strErr = "Error reading wallet database: CPrivKey corrupt";
             return false;
         }
-        if (!pwallet->GetOrCreateLegacyScriptPubKeyMan()->LoadKey(key, vchPubKey))
+        if (!pwallet->GetOrCreateLegacyDataSPKM()->LoadKey(key, vchPubKey))
         {
             strErr = "Error reading wallet database: LegacyScriptPubKeyMan::LoadKey failed";
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/walletdb.cpp]: Done as suggested

File: src/wallet/scriptpubkeyman.h [hunk 10]
```diff
@@ -500,7 +500,6 @@ class LegacyScriptPubKeyMan : public LegacyDataSPKM
 
     /* Set the HD chain model (chain child index counters) and writes it to the database */
     void AddHDChain(const CHDChain& chain);
-    const CHDChain& GetHDChain() const { return m_hd_chain; }
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/scriptpubkeyman.h:455]: Done

File: src/wallet/scriptpubkeyman.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp]: Done as suggested

File: src/wallet/wallet.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (cbergqvist) [reply src/wallet/wallet.cpp:4483]: Thanks, sorry for missing that. Worth an explaining comment like `"Reset require_format since reload_wallet will only be called on non-Berkeley DB sub-wallets from here on."` or something more correct?

File: src/wallet/wallet.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (cbergqvist) [reply src/wallet/wallet.h:967]: `LegacyScriptPubKeyMan` is a `LegacyDataSPKM`, so the comment would be more correct, but it could be corrected later when removing `LegacyScriptPubKeyMan` - unless you plan to rename `LegacyDataSPKM` back into `LegacyScriptPubKeyMan` once the deletion of the latter is complete.

File: src/wallet/scriptpubkeyman.cpp (hunk 2 above)
Comment (cbergqvist) [reply src/wallet/scriptpubkeyman.cpp]: Haven't fully grokked the code yet, so I understand if you'd rather not answer my questions, but if you want:

> I don't think this needs to have any additional explanation. keypool_size only matters for ranged descriptors. Non-HD keys do not make ranged descriptors, so the value here is ignored. 

We're in an HD chain `for`-loop, how is this not involving HD keys?
Surely legacy wallets could be HD wallets.

> The loop below is for an entirely different set of descriptors and is unrelated to this.

I'm referring to the loop over `desc_spks` which is using the pubkeys from the `DescriptorScriptPubKeyMan` that's just been created and comes from `DescriptorScriptPubKeyMan::m_map_script_pub_keys` which is filled out in `DescriptorScriptPubKeyMan::TopUpWithDB()` which is affected by `m_keypool_size`.

Would `m_keypool_size` already always be `0` before this change?

File: src/wallet/wallet.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/wallet.h:967]: I think the meaning is well understood, both before and after `LegacyScriptPubKeyMan` is deleted. `SPKM` is an abbreviation for `ScriptPubKeyMan` anyways.

File: src/wallet/scriptpubkeyman.cpp (hunk 2 above)
Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp]: The line this comment is on is in the one that goes over the non-HD keys. However, I see that you're confused about the similar line in the loop below that iterates the hd chains.

The `keypool_size` here is a runtime value that depends on the startup options. It's is set when the `DescriptorScriptPubKeyman` is instantiated on loading, and the value is never persisted to disk.

The value that does matter is the `range_end` stored in `WalletDescriptor`. This is persisted to disk. When we are going over the hd chains, we set that `range_end` to match the chain counter stored in the `CHDChain` that is being migrated. This results in a descriptor that will produce exactly the same scriptPubKeys as the original hd chain. We are enforcing that the migrated descriptors produce exactly the same scriptPubKeys (no more, no less). Setting `keypool_size` could mean that `TopUp()` would generate *more* scriptPubKeys than we need/expect, which would cause migration to assert and crash.

File: src/wallet/wallet.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/wallet.cpp:4484]: Added a comment.

File: src/wallet/scriptpubkeyman.h [hunk 11]
```diff
@@ -276,31 +276,109 @@ static const std::unordered_set<OutputType> LEGACY_OUTPUT_TYPES {
...
+
+    /** Get the DescriptorScriptPubKeyMans (with private keys) that have the same scriptPubKeys as this LegacyScriptPubKeyMan.
+     * Does not modify this ScriptPubKeyMan. */
+    std::optional<MigrationData> MigrateToDescriptor();
+    /** Delete all the records ofthis LegacyScriptPubKeyMan from disk*/
+    bool DeleteRecords();
+};
+
+// Implements the full legacy wallet behavior
+class LegacyScriptPubKeyMan : public LegacyDataSPKM
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/scriptpubkeyman.h:370]: I don't think it's necessary to have a comment saying such.

[the remark this answers is not in the document]
Comment (cbergqvist) [reply src/wallet/scriptpubkeyman.h:370]: If the code is unable to communicate why it is like it is, it's good to explain the otherwise surprising state of it. It would hopefully limit the amount of questions like my prior one from https://github.com/bitcoin/bitcoin/pull/26596#pullrequestreview-2089208116:
> `LegacyScriptPubKeyMan::NewKeyPool()` only touches data from `LegacyDataSPKM`, should it be moved to that class as public (or protected and then exposed in the subclass?). 

 But I guess you're fairly confident it will be deleted in time for the next release anyway.

File: src/wallet/scriptpubkeyman.cpp (hunk 2 above)
Comment (cbergqvist) [reply src/wallet/scriptpubkeyman.cpp]: Thanks for taking the time to clarify this matter!

**Review by cbergqvist — APPROVED** [review]
ACK a8f8a2d5d0e2328cbe2e569c7e12f653d18f456d

(Ran tests on 31632d18ea122def82820770f14d9aa009726114 but only comments changed since then).
Functional tests including `--extended` passed (a few automatically skipped).
`make check` passed.

**Review by theStack — COMMENTED** [review]
Concept ACK

File: src/wallet/wallet.cpp [hunk 12]
```diff
@@ -4346,11 +4355,24 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
...
         was_loaded = true;
+    } else {
+        // Check if the wallet is BDB
+        const auto& wallet_path = GetWalletPath(wallet_name);
+        if (!wallet_path) {
+            return util::Error{util::ErrorString(wallet_path)};
+        }
+        if (!IsBDBFile(BDBDataFile(*wallet_path))) {
+            return util::Error{_("Error: This wallet is already a descriptor wallet")};
+        }
```
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/wallet.cpp:4394]: It is theoretically possible to have a descriptor wallet using BDB under 2 scenarios, however this is still an unsupported configuration:

1. The descriptor wallet was created on an unreleased version after #16528 but before #19077 (there was a period of about 6 months where this could have happened)
2. The user manipulated the database using external tooling where they created a BDB file and manually inserted descriptor wallet records

File: src/wallet/scriptpubkeyman.cpp [hunk 13]
```diff
@@ -1833,7 +1833,7 @@ std::optional<MigrationData> LegacyScriptPubKeyMan::MigrateToDescriptor()
         WalletDescriptor w_desc(std::move(desc), creation_time, 0, 0, 0);
 
         // Make the DescriptorScriptPubKeyMan and get the scriptPubKeys
-        auto desc_spk_man = std::unique_ptr<DescriptorScriptPubKeyMan>(new DescriptorScriptPubKeyMan(m_storage, w_desc, m_keypool_size));
+        auto desc_spk_man = std::make_unique<DescriptorScriptPubKeyMan>(m_storage, w_desc, /*keypool_size=*/0);
```
Comment (furszy) [inline src/wallet/scriptpubkeyman.cpp:1956]: Not really an issue but this works because `TopUp` sets the descriptor range_end to 1 on the first run when the descriptor is not ranged.
As this would fail if we ever change that, what if we set the range properly in `WalletDescriptor` constructor?
E.g.

```c++
WalletDescriptor w_desc(std::move(desc), creation_time, /*range_start=*/0, /*range_end=*/1, /*next_index=*/0)
```

Also, this change doesn't seem to be related to c653f4fdbfe06 description?

**Review by furszy — COMMENTED** [review]
Just a small comment; you don't have to take it.

I believe this PR could get merged quite fast if you leave 89503710386f52d9162f67fcd707eceffa954faa for a follow-up and make `IsMine` compatible with `LegacyDataSPKM` in this PR.

I'm suggesting this because this PR starts testing the newly introduced BDB reader class and allows people to use it, which is great since we want it polished for the next release. The `IsMine` removal is also nice but I wouldn't miss a deadline for it.

**Review by furszy — COMMENTED** [review]
> * To my understanding, the class `LegacyDataSPKM` is supposed to never write to the database (as `BerkeleyRODatabase` obviously doesn't support that), but `LegacyDataSPKM::CheckDecryptionKey` contains a code-path for that (for a necessary rewrite):
>   https://github.com/achow101/bitcoin/blob/a8f8a2d5d0e2328cbe2e569c7e12f653d18f456d/src/wallet/scriptpubkeyman.cpp#L254-L255
>   Should `CheckDecryptionKey` stay in `LegacyScriptPubKeyMan`?

It isn't really an issue because `CheckDecryptionKey` does not verify the outcome of the write call (which would be "successful" because the read-only bdb class return true for all write operations). But.. for the sake of correctness I see two possible paths to follow:

1) As this write is done exclusively on wallets with no checksum on crypted keys, we could move the upgrade code to the loading code in `walletdb.cpp`; removing `fDecryptionThoroughlyChecked` entirely and requiring such ancient wallets to provide the passphrase during load to perform the upgrade (adding a "passphrase" arg to the `loadwallet` RPC command).

2) Introduce a new `WalletBatch::IsReadOnly()` method to avoid calling the write inside `CheckDecryptionKey`.

The first approach will make the legacy wallet code simpler while the second one is a minimal change.

**Comment by achow101** [comment]
> I believe this PR could get merged quite fast if you leave [8950371](https://github.com/bitcoin/bitcoin/commit/89503710386f52d9162f67fcd707eceffa954faa) for a follow-up and make `IsMine` compatible with `LegacyDataSPKM` in this PR.

Moved to #30328

> The `IsMine` removal is also nice but I wouldn't miss a deadline for it.

I think it needs to be in the next release as it is a critical part of migration. I would prefer to have it in a release that people can use before deleting the legacy wallet.
 
>     * The PR description is outdated, as it still mentions adding `BerkeleyRODatabase` which was already merged in [wallet: Implement independent BDB parser #26606](https://github.com/bitcoin/bitcoin/pull/26606)

Updated

>     * in [c653f4f](https://github.com/bitcoin/bitcoin/commit/c653f4fdbfe0607ad9be27e80ad22c9aee314bb7): typo in commit subject (s/MigrateToLegacy/MigrateToDescriptor/)

Fixed

> 
>     * To my understanding, the class `LegacyDataSPKM` is supposed to never write to the database (as `BerkeleyRODatabase` obviously doesn't support that), but `LegacyDataSPKM::CheckDecryptionKey` contains a code-path for that (for a necessary rewrite):
>       https://github.com/achow101/bitcoin/blob/a8f8a2d5d0e2328cbe2e569c7e12f653d18f456d/src/wallet/scriptpubkeyman.cpp#L254-L255
>       Should `CheckDecryptionKey` stay in `LegacyScriptPubKeyMan`?

No, we should not be migrating a wallet if we cannot decrypt the keys. `BerkeleyRODatabase` always returns true for all write operations so it doesn't matter. Also, I don't think it is useful to invert that and have to have specific bypasses for `BerkeleyRODatabase` in all functions that attempts to write to it.

**Review by furszy — COMMENTED** [review]
Strange CI https://github.com/bitcoin/bitcoin/actions/runs/9649050504/job/26611651920?pr=26596 error.. it doesn't seem to be related. The "salvage" command opens the wallet in `DatabaseFormat::BERKELEY` format.

File: src/wallet/wallet.cpp [hunk 14]
```diff
@@ -4438,6 +4461,8 @@ util::Result<MigrationResult> MigrateLegacyToDescriptor(const std::string& walle
     // both before and after reloading. This ensures the set is complete even if one of the wallets
     // fails to reload.
     std::set<fs::path> wallet_dirs;
+    // Reset options.require_format as options are used by reload_wallet which may reload wallets of any format.
+    options.require_format = std::nullopt;
```
Comment (furszy) [inline src/wallet/wallet.cpp:4484]: In f732495191d346b8:

The reset call needs to be moved above the `reload_wallet` function declaration. Otherwise, the wallet will be reloaded in read-only mode for any failure occurring in-between the `reload_wallet` declaration and this line, such as failures during backup creation or the decryption procedure.

I have written a test to trigger this misbehavior: furszy@c24b673 (feel free to pull it here). The test passes when the fix is cherry-picked on top: furszy@2d5d9a4.

Comment (achow101) [reply src/wallet/wallet.cpp]: Fixed and pulled the test. I decided to just move this into `reload_wallet` directly, it doesn't need to live outside of it.

File: src/wallet/scriptpubkeyman.cpp [hunk 15]
```diff
@@ -784,18 +784,39 @@ bool LegacyScriptPubKeyMan::LoadCScript(const CScript& redeemScript)
...
 
 void LegacyScriptPubKeyMan::LoadScriptMetadata(const CScriptID& script_id, const CKeyMetadata& meta)
 {
     LOCK(cs_KeyStore);
+    LegacyDataSPKM::LoadScriptMetadata(script_id, meta);
     UpdateTimeFirstKey(meta.nCreateTime);
-    m_script_metadata[script_id] = meta;
+}
+
+bool LegacyDataSPKM::AddKeyPubKeyInner(const CKey& key, const CPubKey &pubkey)
```
Comment (furszy) [inline src/wallet/scriptpubkeyman.cpp:813]: nano nit:
```suggestion
bool LegacyDataSPKM::AddKeyPubKeyInner(const CKey& key, const CPubKey& pubkey)
```

File: src/wallet/scriptpubkeyman.cpp [hunk 16]
```diff
@@ -784,18 +784,39 @@ bool LegacyScriptPubKeyMan::LoadCScript(const CScript& redeemScript)
...
+{
+    // This function should only be called during loading of a legacy to be migrated.
+    // As such, the wallet should not be encrypted if this is called.
+    LOCK(cs_KeyStore);
+    assert(!m_storage.HasEncryptionKeys());
```
Comment (furszy) [inline src/wallet/scriptpubkeyman.cpp:818]: I'm not sure this is asserting what you expect. In the loading process, we load the encryption keys after the key records. So the encryption keys map should always be empty and `HasEncryptionKeys` should always return false but not for the reasons stated here.

Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp:818]: Yes, that's what the comment is saying. It is because this function is only called during loading, and because legacy wallet records are loaded first, that this assertion makes sense. After all, to be migrated legacy wallets can still be encrypted.

Comment (furszy) [reply src/wallet/scriptpubkeyman.cpp:818]: I think calling `IsLocked()` would be more precise and make more sense then. It seems confusing to say that the wallet can still be encrypted but expect no encryption keys (we understand it because we know the loading procedure but others will definitely not understand it).

Other than that, what if instead of creating this new `LegacyDataSPKM::AddKeyPubKeyInner` method you place this three lines inside the `LegacyDataSPKM::LoadKey` which is only called during load?

**Comment by furszy** [comment]
Now that the `IsMine` changes were moved to another PR, could reorder the commits just so 0cce574 is introduced before c3373a39a1acc94dbf644103fab2e477d49da671. This will let us keep the IsMine assertions inside `MigrateToDescriptor` for now.

File: src/wallet/scriptpubkeyman.cpp (hunk 15 above)
Comment (cbergqvist) [reply src/wallet/scriptpubkeyman.cpp:813]: This made me cringe when first looking at the code too, but I guess retaining the inconsistency makes it easier to search for `AddKeyPubKeyInner(const CKey& key, const CPubKey &pubkey)` and find all 4 places it exists. Changing it in this PR would arguably add noise (on the other hand - if not now, then when?).

**Review by cbergqvist — COMMENTED** [review]
`git range-diff master a8f8a2d 9700c217`

Agree with @furszy on reordering the commits to make it possible to retain the 3 asserts in `MigrateToDescriptor()`.

File: src/wallet/scriptpubkeyman.cpp (hunk 16 above)
Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp:818]: > Other than that, what if instead of creating this new `LegacyDataSPKM::AddKeyPubKeyInner` method you place this three lines inside the `LegacyDataSPKM::LoadKey` which is only called during load?

We can't do that because `LoadKey` needs to call `LegacyScriptPubKeyMan::AddKeyPubKeyInner` for legacy wallets that are being loaded normally. Since migration cannot do all of the things in `LegacyScriptPubKeyMan::AddKeyPubKeyInner`, we need to have `LegacyDataSPKM::AddKeyPubKeyInner`.

> I think calling `IsLocked()` would be more precise and make more sense then. It seems confusing to say that the wallet can still be encrypted but expect no encryption keys (we understand it because we know the loading procedure but others might not understand it).

Actually, looking at this more closely, your original comment was correct. This function is only reached during legacy migration wallet migration loading for unencrypted keys only. Since the encryption keys are loaded later, this check doesn't actually do anything useful.

**Comment by achow101** [comment]
Reordered the commits and preserved the `IsMine` asserts.

File: src/wallet/scriptpubkeyman.cpp (hunk 15 above)
Comment (cbergqvist) [reply src/wallet/scriptpubkeyman.cpp]: (Now grepping "AddKeyPubKeyInner(const CKey& key, const CPubKey& pubkey)" gives 1 result.
grepping "AddKeyPubKeyInner(const CKey& key, const CPubKey &pubkey)" gives 3 results.

It would be nice to align style within the same class between .CPP/.H).

Comment (achow101) [reply src/wallet/scriptpubkeyman.cpp]: Not going to change existing code, especially code that will be deleted, to match style

Comment (cbergqvist) [reply src/wallet/scriptpubkeyman.cpp]: (Agreed, but then I would skip doing the suggested change in the .CPP file).

**Review by cbergqvist — APPROVED** [review]
ACK 8ce3739edbcf6437bf2695087e0ebe8c633df19b

`git range-diff 9700c217~5..9700c217 8ce3739~5..8ce3739` - not the easiest read, also used GH [compare](https://github.com/bitcoin/bitcoin/compare/9700c21704ddad4be07df44393e39057fa386c56..8ce3739edbcf6437bf2695087e0ebe8c633df19b).

Passed `make check` & `test/functional/test_runner.py` including `p2p_handshake.py` which [failed on CI](https://github.com/bitcoin/bitcoin/pull/26596/checks?check_run_id=26906002421).

Only one non-critical reservation: https://github.com/bitcoin/bitcoin/pull/26596#discussion_r1661582339

**Review by theStack — APPROVED** [review]
Code-review ACK 8ce3739edbcf6437bf2695087e0ebe8c633df19b

File: src/wallet/scriptpubkeyman.cpp [hunk 17]
```diff
@@ -784,18 +784,36 @@ bool LegacyScriptPubKeyMan::LoadCScript(const CScript& redeemScript)
...
+    mapKeyMetadata[keyID] = meta;
+}
+
 void LegacyScriptPubKeyMan::LoadKeyMetadata(const CKeyID& keyID, const CKeyMetadata& meta)
 {
     LOCK(cs_KeyStore);
+    LegacyDataSPKM::LoadKeyMetadata(keyID, meta);
     UpdateTimeFirstKey(meta.nCreateTime);
-    mapKeyMetadata[keyID] = meta;
+}
```
Comment (furszy) [inline src/wallet/scriptpubkeyman.cpp:798]: In 7461d0c006c92:

At this point, just a non-blocking nit but you are locking `cs_KeyStore` twice. On here, and another one inside `LegacyDataSPKM::LoadKeyMetadata`.

Same for `LoadScriptMetadata`.

File: src/wallet/scriptpubkeyman.cpp [hunk 18]
```diff
@@ -784,18 +784,36 @@ bool LegacyScriptPubKeyMan::LoadCScript(const CScript& redeemScript)
...
     LOCK(cs_KeyStore);
+    LegacyDataSPKM::LoadScriptMetadata(script_id, meta);
     UpdateTimeFirstKey(meta.nCreateTime);
-    m_script_metadata[script_id] = meta;
+}
+
+bool LegacyDataSPKM::AddKeyPubKeyInner(const CKey& key, const CPubKey& pubkey)
+{
+    LOCK(cs_KeyStore);
+    return FillableSigningProvider::AddKeyPubKey(key, pubkey);
```
Comment (furszy) [inline src/wallet/scriptpubkeyman.cpp:816]: In https://github.com/bitcoin/bitcoin/commit/7461d0c006c92ede2f2595b79a5509eaf3509fb7:

`FillableSigningProvider::AddKeyPubKey` already locks `cs_Keystore` internally.

**Review by furszy — COMMENTED** [review]
Code review ACK 8ce3739edbcf6437bf2695087e0ebe8c633df19b
